src/main/java/sparksoniq/spark/ml/ApplyTransformerRuntimeIterator.java

```diff
@@ -46,43 +44,74 @@ public ApplyTransformerRuntimeIterator(
         Dataset<Row> inputDataset = getInputDataset(context);
         Item paramMapItem = getParamMapItem(context);
 
-        boolean transformerExpectsFeaturesColParam = RumbleMLCatalog
-            .getTransformerParams(this.transformerShortName, getMetadata())
-            .contains(featuresColParamName);
-
-        boolean isFeaturesColumnGenerated = false;
-
-        if (transformerExpectsFeaturesColParam) {
-            Object featuresColValue = new String[] { featuresColParamDefaultValue };
-            if (paramMapItem.getItemByKey(featuresColParamName) != null) {
-                RumbleMLCatalog.validateTransformerParameterByName(
+        // update input dataset and paramMapItem based on the needs of special params
+        for (String specialParamName : RumbleMLCatalog.specialParams) {
+            if (
+                !RumbleMLCatalog.getTransformerParams(this.transformerShortName, getMetadata())
```

Comment:
The code on estimator and transformer sides are %99 identical with only changes being the called getter methods from the RumbleMLCatalog. When the system has matured, I believe that estimators and transformers can be merged together to have a unified API with shared maps and getter methods. However, I deliberately did not take this approach in the initial implementation as before said maturity is reached, I wanted to keep the code flexible enough to accommodate unforeseen peculiarities of Estimators and transformers.